src/services/completions.ts

```diff
@@ -335,6 +325,42 @@ namespace ts.Completions {
         }
     }
 
+    function getJsxClosingTagCompletion(location: Node | undefined, sourceFile: SourceFile): CompletionInfo | undefined {
+        // We wanna walk up the tree till either we find a JSX closing element, JSX opening element or quit at end
+        const jsxClosingElement = findAncestor(location, node => isJsxOpeningElement(node)
+                ? "quit"
+                : isJsxClosingElement(node)) as JsxClosingElement | undefined;

```

Comment:
> The only thing you’ll ever hit where you need to keep going up is a PropertyAccessExpression. Just continue as long as you have an Identifier or a PropertyAccessExpression.

This is what I mean. Do this:

```suggestion
        const jsxClosingElement = findAncestor(location, node => {
            switch (node.kind) {
                case SyntaxKind.JsxClosingElement:
                    return true;
                case SyntaxKind.SlashToken:
                case SyntaxKind.GreaterThanToken:
                case SyntaxKind.Identifier:
                case SyntaxKind.PropertyAccessExpression:
                    return false;
                default:
                    return "quit";
            }
        }) as JsxClosingElement | undefined;

```

Reply:
Love it man, really like the pattern matching here well done!

Reply:
The advantage is if we see we’re in something like a variable declaration, or a call expression, or almost anything else, we _know_ we can’t be in a closing tag and we quit. We don’t have to keep going up 10 parents before quitting. But at the same time, we could traverse through more than 10 parents if there _was_ a super super long property access expression in a closing tag.

Reply:
Totally agree. I have to study more the element tree and the hierarchy to see how it's constructed. I can't make good calls when I don't understand how things work...

Reply:
An [AST viewer](https://ts-ast-viewer.com/#) is useful, especially while still learning, if you’re not already referencing one.

Reply:
Yup thats better. I has completely misunderstood how the AST is constructed.